### ISSUE TITLE
Implement Java 8 type argument inference

### ISSUE THREAD

**Issue by smillst** [typetools/checker-framework#979]
The Checker Framework should implement [Java 8 type inference](https://docs.oracle.com/javase/specs/jls/se8/html/jls-18.html) for method type arguments, new class trees that use diamonds, and raw types.  (Currently, the Checker Framework implements Java 7 type inference.) 
 This involves completely replacing/rewriting the implementation in the `framework.utils.typeinference` package.  Also, the code that adds type arguments to raw types (`AnnotatedDeclaredType#getTypeArguments`) and the code that infers type arguments for diamonds (`AnnotatedTypeFactory#fromNewClass`) should be changed to use type argument inference.

The Java 8 algorithm specified in the JLS cannot be used as is.  It must be adapted to use with annotated types.  In particular, annotated types complicate the JLS  algorithm in following ways. 

1.  In the JLS, if there is a subtyping constraint against a null type, the constraint is reduced to "true".[[JLS  #18.2.3]](https://docs.oracle.com/javase/specs/jls/se8/html/jls-18.html#jls-18.2.3)  However, if null isn't annotation with the bottom qualifier, then the constraint should be reduced to new constraint where the primary annotation of the type must be a subtype of the annotation on null.

2.  If a use of a type variable has a primary annotation, then constraints against that use should not include that annotation.  In other words, that constraint should be against the Java type rather than the annotated type.  But if the type system has a multigraph qualifier hierarchy, then there should still be a constraint against an annotated type for some of the hierarchies.   (See examples in checker/tests/nullness/generics/MethodTypeVars7.java.) 

3. A class can be marked with `@Covariant` to signify that one or more of its type parameters can be treated covariantly.  Covariant type parameters should generate subtyping constraints rather than equality constraints.

4. Method postconditions. See #1345. 

Also, once Java 8 inference is implemented, the field `AnnotatedWildcardType#uninferredTypeArgument` should be removed and the `-AconservativeUninferredTypeArguments` option should also be removed.
In the meanwhile, if you are concerned that the Checker Framework is suffering false negatives due to this issue, use `-AconservativeUninferredTypeArguments`.  A user [reported](https://github.com/typetools/checker-framework/issues/2235#issuecomment-452862815) that this required editing 0.5% of the lines of code in his project.

The following issues were closed in favor of this issue: #402, #404,  #531, #953, #980, #1032.  The test cases from those issues are now in framework/tests/all-systems/java8inferrence/ and checker/tests/nullness/java8inferrence/.

**Comment by wmdietlGC** [comment]
As people are adopting Stream-based APIs, I'm running into this issue very frequently.

I have a patch that instead of raising errors like:

```
Issue404.java:17: error: [return.type.incompatible] incompatible types in return.
        return strings.stream().map(String::trim).collect(Collectors.toSet());
                                                         ^
  found   : @Initialized @NonNull Set<? extends @Initialized @Nullable Object>
  required: @Initialized @NonNull Set<@Initialized @NonNull String>
```

allows uninferred wildcards as subtypes. The Nullness Checker then successfully compiles all examples in `tests/all-systems/java8inference`.
This is unsound, but there is currently no test case that would require an error. We should add one to highlight the problem.

The current error message isn't useful and very frustrating.
I would suggest making the default behavior unsound (it hurts me to say that...) and adding an option to use the current conservative behavior.

Let me know what you think.

**Comment by smillst** [comment]
https://github.com/typetools/checker-framework/issues/979#issuecomment-279578221 was implemented via bfeee16.

**Comment by kelloggm** [comment]
I encountered a problem that I believe is related to the type inference algorithm. This example, adapted from real code, throws a very unhelpful error that doesn't even point to the offending code (we found it via delta debugging in a much larger codebase):
```java
public class LambdasAndGenericStaticMethods {
    private static TwoTypeArguments throwsTheError() {
        return lambdaArg -> lambdaArg
                .compose(genericStaticMethod());
    }

    private static <T> TwoTypeArguments<T, T> genericStaticMethod() {
        return null;
    }

    private interface TwoTypeArguments<T, R> {
        OneTypeArgument<R> apply(AbstractClassWithOneTypeArgument<T> arg);
    }

    private interface OneTypeArgument<T> {
        void method(T t);
    }

    private abstract class AbstractClassWithOneTypeArgument<T> implements OneTypeArgument<T> {
        public final <R> AbstractClassWithOneTypeArgument<R> compose(TwoTypeArguments<? super T, ? extends R> composer) {
            return null;
        }
    }
}
```
As far as we can tell, all of these elements need to be present for the crash to occur. The resulting error looks like this:
```
error: Unexpected F2A + Combination:
  subtype=? extends Object
  supertype= null
  constraints=[

  ]; invoke the compiler with -AprintErrorStack to see the stack trace.
error: Unexpected F2A + Combination:
  subtype=? extends @Initialized @NonNull Object
  supertype=null
  constraints=[

  ]; invoke the compiler with -AprintErrorStack to see the stack trace.
```
Note that there is no way, from this error, to trace it back to the offending line. A workaround is to add explicit type arguments throughout the function `throwsTheError`. Then the type inference algorithm is able to proceed normally. I found this while running the nullness checker, but it's not nullness checker exclusive.

**Comment by smillst** [comment]
@kelloggm It's easier to track crashes from type argument inference in separate issues, so I opened I new on #1512 for this crash.  (I'll eventually delete this comment and the one above.)

**Comment by kelloggm** [comment]
Sorry, I thought this was the tracking issue since it mentions several issues that were closed in favor of this one.

Thanks for moving it.

**Comment by msridhar** [comment]
Is there any rough estimate on when this issue might get addressed?  We have an internal checker for stream-related code and we're running into this limitation frequently.  I understand it must be a significant amount of work to fix, though, and it may take a while.  Just wondering how far we should attempt to go with workarounds 😄

**Comment by mernst** [comment]
@msridhar , Suzanne has been working on this issue since the summer, spending 2-3 days per week.  So, she has devoted 1-3 months to it so far.  It will probably be months before we can close the issue, unless by a miracle no other distractions come up for her.  This suggests that you may want to put workarounds in place.  Sorry about that.  We are looking forward to removing a number of workarounds in the Checker Framework proper, once the issue is fixed.

**Comment by justmin** [comment]
Hello, 
Optional::map doesn't accept @Nullable #1633 could be fixed (or workarounded?) quickly.
Could you please include this changes in the next release?
We use Stream::map very often and need for workarounding it is the main argument against using CF, at least in my project

**Comment by bmhm** [comment]
To fix #402 (comparators with nullsFirst/Last), you can use this stub:

```
import org.checkerframework.checker.nullness.qual.NonNull;

package java.util;

public interface Comparator<T> {
  public static <@NonNull T> Comparator<T> nullsFirst(Comparator<? super T> comparator);
}
```

I hope this helps someone intil it is finally resolved and added to the shipped annotated jdk8.

**Comment by wmdietlGC** [comment]
Another example, related to lambdas:

```java
import org.checkerframework.checker.nullness.qual.Nullable;

interface Box<S> {}

interface Getter<T> {
  T get();
}

interface Runner {
  <U> Box<U> run(Getter<? extends Box<? extends U>> g);
}

abstract class Demo {
  Box<@Nullable String> borked(Runner r) {
    return r.run( () -> get() );
  }

  Box<@Nullable String> works(Runner r) {
    return r.run(
        (Getter<Box<@Nullable String>>) (() -> get())
      );
  }

  abstract Box<@Nullable String> get(); 
}
```

This gives the following unexpected error in `borked`:

```
Error: [return.type.incompatible] incompatible types in return.
   type of expression: @Initialized @NonNull Box<@Initialized @Nullable String>
   method return type: @Initialized @NonNull Box<? extends @Initialized @NonNull String>
```

It is a bit surprising to see `@NonNull` String as the inferred type.

**Comment by wmdietlGC** [comment]
The following code:

```java
import org.checkerframework.checker.nullness.qual.Nullable;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collector;
import java.util.Map;

class Demo {
  static <T, R, A> void use(@Nullable Collector<? super T, A, R> collector) { }

  static <T, K, V> @Nullable Collector<T, ?, Map<K,V>> calc(Function<? super T, ? extends K> keyFunction,
    Function<? super T, ? extends V> valueFunction,
    BiFunction<V, V, V> mergeFunction) { return null; }

  void foo(Function<Long, Float> f) {
    // No error when using a lambda
    use(calc(f, f, (a, b) -> Math.max(a, b)));
    // Error when using a method reference
    use(calc(f, f, Math::max));
  }
}
```

gives the following errors with the Nullness Checker:

```
Demo.java:18: error: [methodref.param.invalid] Incompatible parameter type for a
    use(calc(f, f, Math::max));
                   ^
  Method
    @Initialized @NonNull float max(@Initialized @NonNull float p0, @Initialized @NonNull float p1) in java.lang.Math
  is not a valid method reference for
    /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object apply(@Initialized @NonNull BiFunction</*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object, /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object, /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object> this, /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object p0, /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object p1) in java.util.function.BiFunction
  found   : @Initialized @NonNull float
  required: /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object
Demo.java:18: error: [methodref.param.invalid] Incompatible parameter type for b
    use(calc(f, f, Math::max));
                   ^
  Method
    @Initialized @NonNull float max(@Initialized @NonNull float p0, @Initialized @NonNull float p1) in java.lang.Math
  is not a valid method reference for
    /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object apply(@Initialized @NonNull BiFunction</*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object, /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object, /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object> this, /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object p0, /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object p1) in java.util.function.BiFunction
  found   : @Initialized @NonNull float
  required: /*INFERENCE FAILED for:*/ ? extends @Initialized @Nullable Object
2 errors
```

Note how the equivalent, but more verbose, lambda expression works without errors.

I would have expected these errors to be suppressed by default, because the types contain `INFERENCE FAILED`. Maybe that suppression is implemented for lambdas but not for method references?
Maybe this can be fixed independently of the larger fix for #979?

**Comment by msridhar** [comment]
@wmdietlGC #3044 was the previous attempt to suppress these types of warnings by default, I think.  #3568 shows another example where errors are still reported with `INFERENCE FAILED`.  Maybe open a new issue for the case above?  IMO all such errors should be suppressed by default.

**Comment by wmdietlGC** [comment]
@msridhar Thanks for the cross-references! I've opened #3764.
